compiler/rustc_infer/src/infer/outlives/verify.rs

```diff
@@ -82,27 +83,39 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
         debug!("param_bound(param_ty={:?})", param_ty);
 
         // Start with anything like `T: 'a` we can scrape from the
-        // environment
-        let param_bounds = self
-            .declared_generic_bounds_from_env(GenericKind::Param(param_ty))
-            .into_iter()
-            .map(|outlives| outlives.1);
+        // environment. If the environment contains something like
+        // `for<'a> T: 'a`, then we know that `T` outlives everything.
+        let declared_bounds_from_env = self.declared_generic_bounds_from_env(param_ty);
+        let mut param_bounds = vec![];
+        for declared_bound in declared_bounds_from_env {
+            let bound_region = declared_bound.map_bound(|outlives| outlives.1);
+            if let Some(region) = bound_region.no_bound_vars() {
+                // This is `T: 'a` for some free region `'a`.
+                param_bounds.push(VerifyBound::OutlivedBy(region));
+            } else {
+                // This is `for<'a> T: 'a`. This means that `T` outlives everything! All done here.
+                return VerifyBound::AllBounds(vec![]);
+            }
```

Comment:
I wonder if it's better to dicharge `for<'a> T: 'a` as `T: 'static` earlier when elaborating implied bounds.

Currently `for<'a> (T,): 'a` doesn't imply itself, which is surprising since we now allow more complex projection bounds to imply themselve `for<'a> <&'a () as Tr<'a>>::Assoc: 'a`.

Reply:
My expectation is that we would make `for<'a> T: 'a` an implied bound of `for<'a> (T,): 'a`, just as we currently make `T: 'a` an elaboration of `(T,): 'a`.  I guess we skip that for higher-ranked things?

Relevant code:

https://github.com/rust-lang/rust/blob/6499c5e7fc173a3f55b7a3bd1e6a50e9edef782d/compiler/rustc_infer/src/traits/util.rs#L210-L216